File: 516-longest-palindromic-subsequence/longest-palindromic-subsequence.java
class Solution {
    public int longestPalindromeSubseq(String s) {
        int n=s.length();
        int [][]dp = new int[n][n];
        return solve(0,n-1,s,dp);
    }
    public int solve(int i,int j,String s,int[][]dp){
        if(i>j) return 0;
        if(i==j) return 1;
        if(dp[i][j]!=0) return dp[i][j];
        if(s.charAt(i)==s.charAt(j)){
            dp[i][j]= 2+solve(i+1,j-1,s,dp);
        }else{
            dp[i][j]= Math.max(solve(i+1,j,s,dp),solve(i,j-1,s,dp));
        }
        return dp[i][j];
    }
}

File: 516-longest-palindromic-subsequence/longest-palindromic-subsequence.py
class Solution:
    def longestPalindromeSubseq(self, s: str) -> int:
        n = len(s)
        dp = [[-1] * n for _ in range(n)]
        def solve(i, j):
            if i > j :
                return 0

            if i == j :
                return 1

            if dp[i][j] != -1:
                return dp[i][j]

            if s[i] == s[j] :
                dp[i][j] = 2 + solve(i+1, j-1)

            else:
                dp[i][j] = max(solve(i+1,j), solve(i, j-1))


            return dp[i][j]


        return solve(0, n-1)

        
        